KIPs/kip-146.md

```diff
@@ -0,0 +1,157 @@
+
+Parameters:
+
+- `validators`: deterministically sorted list of validators.
+- `round`: the target consensus round.
+- `seed`: `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` (from [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md)) of the previous block is used (i.e., previous to the target block).
+
+```py
+def shuffle_validators_KIP146(validators: list[int], round: int, seed: int):
+    tmp = validators[:]
+    ret = []
+
+    random.seed(seed)
+    random.shuffle(tmp)
+    ret.append(tmp[0])
+
+    for i in range(round):
+        tmp = tmp[i + 1 :]
+        if len(tmp) == 0:
+            break
+        random.shuffle(tmp)
+        ret.append(tmp[0])
+    return ret
+```
+
+### Proposer selection
+
+For calculating the proposer of `FORK_BLOCK` and afterward, the new logic must be used.
+This only affects the chain whose proposer selection policy is set to `WeightedRandom`.
+
+This design allows a certain validator to be selected as the proposer for consecutive blocks, while guaranteeing another proposer to be selected in case of round change.
+
+Parameters:
+
+- `validators`: the deterministically sorted list of validators.
+- `round`: the target consensus round.
+- `seed`: `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` (from [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md)) of the previous block is used (i.e., previous to the target block).
+
+```py
+def proposer_selector(validators, round, seed):
+    return shuffle_validators_KIP146(validators, round, seed)[round % len(validators)]
+```
+
+### Committee selection
+
+For calculating the committee of `FORK_BLOCK` and afterward, the new logic must be used.
+This only affects the chain whose proposer selection policy is set to `WeightedRandom`.
+
+This design guarantees that the given proposer address is always in the committee.
+
+Parameters:
+
+- `validators`: the deterministically sorted list of validators.
+- `commitee_size`: the value of `istanbul.sub` in ChainConfig.
+- `seed`: `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` (from [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md)) of the previous block is used (i.e., previous to the target block).
+- `round`: the target consensus round.
+
+```py
+def select_committee_KIP146(validators, committee_size, seed, round):
+    shuffled = shuffle_validators_KIP146(validators, round, seed)
+    return (shuffled[round:] + shuffled[:round])[:committee_size]
+
+def compute_committee(validators, committee_size, proposer, round, seed):
+    committee = select_committee_KIP146(validators, committee_size, seed, round)
+    if proposer not in committee:
```

Comment:
It seems like the first element of `committee` would be the proposer. Can you explain why we need this statement?

Reply:
This is a legacy from the current Klaytn code. Maybe we can remove this logic after the hardfork.

In detail, [SubListWithProposer](https://github.com/klaytn/klaytn/blob/v1.10.2/consensus/istanbul/validator/weighted.go#L336) takes a "proposer" as a argument. It will guarantee that the given proposer is always included.

Consider the following example: when `valset = [v1, v2, v3, v4, v5]` and `valSet.GetProposer() = v2`, `SubListWithProposer(parentHash, v1, view)` can return a random committee [v1, v4, v5], which shows that
- given proposer v1 is included
- the "correct" proposer v2 may not be included

In Klaytn code, it seems like the callers of `SubListWithProposer` will always pass the correct proposer, but I'm not sure about that.

Reply:
Okay. I see. What about adding a log in the actual code to check whether it really happens or not?
Not here in the spec.
Thanks.

Reply:
I concluded that `SubListWithProposer` always takes proposer address as an argument, so I clarified "The proposer must be selected from committee" in spec.